src/bun.js/api/sockets.classes.ts

```diff
@@ -0,0 +1,135 @@
+import { define } from "../scripts/class-definitions";
+
+function generate(ssl) {
+  return define({
+    name: ssl ? "TCPSocket" : "TLSSocket",
+    JSType: "0b11101110",
+    proto: {
+      write: {
+        fn: "write",
+        length: 3,
+      },
+      end: {
```

Comment:
Great questions.

- `shutdown`: it stops new reads/writes to the socket. This corresponds to https://man7.org/linux/man-pages/man2/shutdown.2.html
- `end`: `write()` but it sends a TCP FIN packet. It lets you write something as the final chunk of data to send. The choice of `end` as the name is for consistentcy with Node
- `close` is end() but without any data to send and it more immediately terminates it

Reply:
Gotcha. I ask because it might be confusing to have 3 different methods for "close this thing". I suppose Node does have these [APIs](https://nodejs.org/api/net.html), but to anyone who doesn't know syscalls (which is most people), I think we should consider making it more simple.

If I understand correctly, it could look something like this:
```ts
interface Socket {
  // Sends a FIN packet with no data.
  // (If you didn't read the docs, you'd probably pick this)
  close(): void;
  // Sends a FIN packet with data.
  close(data: ArrayBuffer): number;
  // Stops I/O activity and does not send a FIN packet.
  closeNow(): void;
}
```

It's easier for IDE auto-complete and discoverability if it's just 1 name. If there are 3 different names it's confusing and people will likely pick the wrong thing.

Reply:
One thing to keep in mind with Socket API versus other APIs

This is very low-level. Nobody is using a socket API for fun. They're using it to implement Redis, Postgres, MySQL etc. Leaking obscure details about system calls used here is kind of okay.

That being said, I agree we should combine end or close. I think we should do `end` to be more consistent with Node + we use `end` for `ArrayBufferSink`